File: python/ode_adapter/hl7v2.py
"""Minimal HL7 v2 handling for the 360X workflow envelope.

360X carries its workflow state in HL7 v2 messages alongside C-CDA documents:
  - OMG^O19  General Clinical Order  -> PCC-55 Referral Request, PCC-57 Outcome,
                                         PCC-59 Interim Consultation Note (with doc)
  - OSU^O51  Order Status Update     -> PCC-56 Referral Request Status (accept/decline),
                                         PCC-58 Referral Cancellation
  - SIU^S12/S26 Scheduling           -> PCC-60 Appointment, PCC-61 No-Show

This module implements only what the reference flow needs: a tiny pipe-delimited
parser and builders for the messages above. A production adapter should use a
full v2 library and validate against the 360X transaction tables. Field bindings
below are illustrative and MUST be validated against IHE PCC 360X Vol. 2.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone


def _ts() -> str:
    return datetime.now(timezone.utc).strftime("%Y%m%d%H%M%S")


# Transaction -> v2 message type (validate against 360X Vol. 2 transaction tables)
TX_MESSAGE_TYPE = {
    "PCC-55": "OMG^O19",   # Referral Request
    "PCC-56": "OSU^O51",   # Referral Request Status (accept / decline)
    "PCC-57": "OMG^O19",   # Referral Outcome (with consultation note)
    "PCC-58": "OSU^O51",   # Referral Cancellation
    "PCC-59": "OMG^O19",   # Interim Consultation Note (with note)
    "PCC-60": "SIU^S12",   # Appointment Notification
    "PCC-61": "SIU^S26",   # No-Show Notification
}


@dataclass
class V2Message:
    """A parsed-enough representation of a 360X v2 message."""
    message_type: str                 # e.g. "OSU^O51"
    control_id: str
    referral_id: str | None = None    # placer/referral identifier (the loop key)
    order_status: str | None = None   # for OSU^O51: e.g. "IP" accepted, "CA" cancelled
    fields: dict = field(default_factory=dict)
    raw: str = ""


def parse(raw: str) -> V2Message:
    """Parse the handful of segments the reference flow inspects."""
    segs = [s for s in raw.replace("\n", "\r").split("\r") if s.strip()]
    msg = V2Message(message_type="", control_id="", raw=raw)
    for seg in segs:
        f = seg.split("|")
        sid = f[0]
        if sid == "MSH":
            msg.message_type = f[8] if len(f) > 8 else ""
            msg.control_id = f[9] if len(f) > 9 else ""
        elif sid == "ORC":
            # ORC-2 placer order number used as referral id; ORC-5 order status.
            # Reply content (degraded 360X side): ORC-12 accepting provider,
            # ORC-15 expected-by date, ORC-16 coded status reason.
            if len(f) > 2 and f[2]:
                msg.referral_id = f[2]
            if len(f) > 5 and f[5]:
                msg.order_status = f[5]
            if len(f) > 12 and f[12]:
                msg.fields["accepting_provider"] = f[12]
            if len(f) > 15 and f[15]:
                msg.fields["period_end"] = f[15]
            if len(f) > 16 and f[16]:
                msg.fields["status_reason"] = f[16]
        elif sid == "NTE":
            # NTE-3 comment carries the free-text reply note / reason (degraded).
            if len(f) > 3 and f[3]:
                msg.fields["note"] = f[3]
        elif sid == "ZRF" and len(f) > 1:  # 360X referral-id convention (illustrative)
            msg.referral_id = msg.referral_id or f[1]
        elif sid == "SCH":
            # SIU appointment timing: SCH-11 start, SCH-12 end (illustrative binding).
            if len(f) > 11 and f[11]:
                msg.fields["appointment_start"] = f[11]
            if len(f) > 12 and f[12]:
                msg.fields["appointment_end"] = f[12]
    return msg


def _set_idx(seg: list[str], idx: int, value: str) -> None:
    while len(seg) <= idx:
        seg.append("")
    seg[idx] = value


def build(message_type: str, referral_id: str, order_status: str | None = None,
          control_id: str | None = None, appointment_start: str | None = None,
          *, appointment_end: str | None = None, note: str | None = None,
          accepting_provider: str | None = None, status_reason: str | None = None,
          period_end: str | None = None) -> str:
    """Build a minimal v2 message for the given 360X transaction.

    For SIU scheduling messages (PCC-60/61) an `appointment_start`/`appointment_end`
    (HL7 v2 ts or ISO-8601) is carried in an SCH segment so the inbound bridge can
    build an Appointment.

    Reply content (the degraded 360X side of the crosswalk):
      `accepting_provider` -> ORC-12, `period_end` -> ORC-15, `status_reason` ->
      ORC-16, and `note` -> an NTE segment. Field bindings are illustrative —
      validate against 360X Vol. 2.
    """
    control_id = control_id or f"ADP{_ts()}"
    msh = "|".join([
        "MSH", "^~\\&", "OHIA-360ODE-ADAPTER", "OHIA",
        "EHR", "ORG", _ts(), "", message_type, control_id, "P", "2.5.1",
    ])
    segs = [msh]
    orc_status = order_status or ""
    # ORC-1 order control; map by message
    order_control = "SC" if message_type.startswith("OSU") else "NW"
    orc = ["ORC", order_control, referral_id, "", "", orc_status]
    if accepting_provider:
        _set_idx(orc, 12, accepting_provider)
    if period_end:
        _set_idx(orc, 15, period_end)
    if status_reason:
        _set_idx(orc, 16, status_reason)
    segs.append("|".join(orc))
    if message_type.startswith("SIU"):
        sch = ["SCH"] + [""] * 12
        sch[11] = appointment_start or ""
        sch[12] = appointment_end or ""
        segs.append("|".join(sch))
    if note:
        segs.append("|".join(["NTE", "1", "", note]))
    return "\r".join(segs) + "\r"

File: dotnet/src/Ode.Adapter/Hl7v2.cs
namespace Ode.Adapter;

/// <summary>A parsed-enough 360X v2 message. Mirrors hl7v2.V2Message.</summary>
public sealed class V2Message
{
    public string MessageType { get; set; } = "";
    public string ControlId { get; set; } = "";
    public string? ReferralId { get; set; }
    public string? OrderStatus { get; set; }
    // Reply content parsed off the degraded 360X side (accepting provider, period end,
    // status reason, note, appointment start/end). Keys mirror hl7v2.py's V2Message.fields.
    public Dictionary<string, string> Fields { get; } = new();
    public string Raw { get; set; } = "";
}

/// <summary>Minimal HL7 v2 handling for the 360X workflow envelope. Mirrors hl7v2.py.</summary>
public static class Hl7v2
{
    /// <summary>Transaction -> v2 message type (validate against 360X Vol. 2).</summary>
    public static readonly IReadOnlyDictionary<string, string> TxMessageType =
        new Dictionary<string, string>
        {
            ["PCC-55"] = "OMG^O19",
            ["PCC-56"] = "OSU^O51",
            ["PCC-57"] = "OMG^O19",
            ["PCC-58"] = "OSU^O51",
            ["PCC-59"] = "OMG^O19",
            ["PCC-60"] = "SIU^S12",
            ["PCC-61"] = "SIU^S26",
        };

    private static string Ts() => DateTime.UtcNow.ToString("yyyyMMddHHmmss");

    public static V2Message Parse(string raw)
    {
        var msg = new V2Message { Raw = raw };
        var segs = raw.Replace("\n", "\r").Split('\r', StringSplitOptions.RemoveEmptyEntries);
        foreach (var seg in segs)
        {
            if (seg.Trim().Length == 0) continue;
            var f = seg.Split('|');
            var sid = f[0];
            if (sid == "MSH")
            {
                msg.MessageType = f.Length > 8 ? f[8] : "";
                msg.ControlId = f.Length > 9 ? f[9] : "";
            }
            else if (sid == "ORC")
            {
                // ORC-2 placer order number = referral id; ORC-5 order status.
                // Reply content: ORC-12 accepting provider, ORC-15 expected-by date,
                // ORC-16 coded status reason.
                if (f.Length > 2 && f[2].Length > 0) msg.ReferralId = f[2];
                if (f.Length > 5 && f[5].Length > 0) msg.OrderStatus = f[5];
                if (f.Length > 12 && f[12].Length > 0) msg.Fields["accepting_provider"] = f[12];
                if (f.Length > 15 && f[15].Length > 0) msg.Fields["period_end"] = f[15];
                if (f.Length > 16 && f[16].Length > 0) msg.Fields["status_reason"] = f[16];
            }
            else if (sid == "NTE")
            {
                if (f.Length > 3 && f[3].Length > 0) msg.Fields["note"] = f[3];
            }
            else if (sid == "ZRF" && f.Length > 1)
            {
                msg.ReferralId ??= f[1];
            }
            else if (sid == "SCH")
            {
                if (f.Length > 11 && f[11].Length > 0) msg.Fields["appointment_start"] = f[11];
                if (f.Length > 12 && f[12].Length > 0) msg.Fields["appointment_end"] = f[12];
            }
        }
        return msg;
    }

    private static void SetIdx(List<string> seg, int idx, string value)
    {
        while (seg.Count <= idx) seg.Add("");
        seg[idx] = value;
    }

    public static string Build(string messageType, string referralId,
                               string? orderStatus = null, string? controlId = null,
                               string? appointmentStart = null, string? appointmentEnd = null,
                               string? note = null, string? acceptingProvider = null,
                               string? statusReason = null, string? periodEnd = null)
    {
        controlId ??= $"ADP{Ts()}";
        var msh = string.Join("|", new[]
        {
            "MSH", "^~\\&", "OHIA-360ODE-ADAPTER", "OHIA",
            "EHR", "ORG", Ts(), "", messageType, controlId, "P", "2.5.1",
        });
        var segs = new List<string> { msh };
        var orderControl = messageType.StartsWith("OSU") ? "SC" : "NW";
        var orc = new List<string> { "ORC", orderControl, referralId, "", "", orderStatus ?? "" };
        if (!string.IsNullOrEmpty(acceptingProvider)) SetIdx(orc, 12, acceptingProvider);
        if (!string.IsNullOrEmpty(periodEnd)) SetIdx(orc, 15, periodEnd);
        if (!string.IsNullOrEmpty(statusReason)) SetIdx(orc, 16, statusReason);
        segs.Add(string.Join("|", orc));
        if (messageType.StartsWith("SIU"))
        {
            var sch = new List<string> { "SCH" };
            for (int i = 0; i < 12; i++) sch.Add("");
            sch[11] = appointmentStart ?? "";
            sch[12] = appointmentEnd ?? "";
            segs.Add(string.Join("|", sch));
        }
        if (!string.IsNullOrEmpty(note)) segs.Add(string.Join("|", new[] { "NTE", "1", "", note }));
        return string.Join("\r", segs) + "\r";
    }
}
